### PR TITLE
Factor out a GATKBaseTest for separate test resources from test utilities in BaseTest

### LINKED ISSUE #2125: Remove GenomeLocParser from BaseTest
BaseTest instantiates a GenomeLocParser that uses one of our test references.  This causes BaseTest to be unusable by external projects that don't have the file included.  We should either package the reference file in the jar or we should remove the GenomeLocParser from BaseTest and instantiate it in the classes that actually use it.

### LINKED ISSUE #3029: For projects built on GATK framework, automated testing is broken
This applies to projects that import the GATK jar as part of the build process, but are not part of the GATK itself.

All unit and integration tests are (by default) broken, since the BaseTest class requires the mini fasta, even when it should not be required.  This causes breakage, since a project built on the GATK should not be expected to have that file at the exact correct place in the filesystem.

The tests do not even start.

### PR DESCRIPTION
**Pull request by magicDGS**
Changes to the testing framework to remove references to the test resources, keeping them into the src/test package. This changes include:

* Factor out a `GATKBaseTest` for separate test resources from test utilities in `BaseTest`
* Remove duplicated `CleanSamIntegrationTest`
* Repackage `CommandLineProgramTest` to be in the test sources, and use it's interface in testers
* Move some testers to the src/test package because they are tool-specific (added TODO to other ones that aren't that clear)
* Refactor `TargetsToolsTestUtils` to use a provided reference

Closes #3029
Closes #2125

### CODE REVIEW

**Comment by magicDGS** [comment]
Here is a new PR for the issues related with the test code. Back to you @droazen.

**Comment by magicDGS** [comment]
@droazen, I updated the PR message to describe the changes, because I tried to get all the code referring to test resources packaged in src/test. Now it should pass the tests too.

**Comment by jamesemery** [comment]
@magicDGS Could you rebase this and update newer references to baseTest that have been added since you created this branch?

**Comment by magicDGS** [comment]
@jamesemery - I think that the rebase is done. I'd like to have this in as soon as it can be, to avoid the extra-work of rebasing due to new tests or refactoring of them.... Thank you in advance!

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java [hunk 1]
```diff
@@ -168,25 +90,6 @@ protected static String getNonNullEnvironmentVariable(String envVarName) {
...
-        return intervalStringsToGenomeLocs(Arrays.asList(intervals));
-    }
-
-    protected List<GenomeLoc> intervalStringsToGenomeLocs( List<String> intervals ) {
-        List<GenomeLoc> locs = new ArrayList<>();
-        for (String interval: intervals)
-            locs.add(hg19GenomeLocParser.parseGenomeLoc(interval));
-        return Collections.unmodifiableList(locs);
-    }
-
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: For purity sake I think getToolTestDataDir() should be left abstract in BaseTest and instantiated in GATKBaseTest.

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java [hunk 2]
```diff
@@ -15,25 +14,21 @@
...
-    private TargetsToolsTestUtils(){} //don't instantiate this
-
-    /**
-     * Returns a {@link File} pointing to the directory that contains the test data.
-     * @return never {@code null}.
-     */
-    public static File getTestDataDir(){
-        return new File(CommandLineProgramTest.getTestDataDir(),"exome");
+    /** Initialize the reference file and dictionary to use for creating intervals. */
+    public TargetsToolsTestUtils(final File referenceFile){
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: It should not be the case that you are ever instantiating a utils class.

Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: I think if you are trying to avoid calling = new File(getTestDataDir(),"test_reference.fasta"), create a getter which farms out to GATKBaseTest getTestDataDir() to get these files and leave it to external tool authors to avoid using these files.

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java [hunk 3]
```diff
@@ -77,7 +72,8 @@ public static SimpleInterval createOverEntireContig(final String contig) {
      * @return never {@code null}.
      * @throws UserException if there was some problem when creating the location.
      */
-    public static SimpleInterval createInterval(final String contig, final int start) {
+    public SimpleInterval createInterval(final String contig, final int start) {
+        // TODO: should this really be createInterval(contig, start, start) instead of using the constructor supplied here?
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: I agree with you, this method is kindof pointless, it seems to only be called in one place and furthermore it seems to be used in one place, and given the the other overload it seems confusing to just use SimpleInterval

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/CleanSamIntegrationTest.java [hunk 4]
```diff
@@ -1,58 +0,0 @@
...
-import picard.sam.CleanSam;
-import org.testng.Assert;
-
-import java.io.PrintWriter;
-import java.util.*;
-
-/**
- * This class is the extension of the SamFileTester to test CleanSam with SAM files generated on the fly.
- */
-public final class CleanSamIntegrationTest extends SamFileTester {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/test/testers/CleanSamIntegrationTest.java]: Nice catch on the duplicated code

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java [hunk 5]
```diff
@@ -50,7 +53,7 @@ public File getMetricsFile() {
     }
 
     @Override
-    public String getTestedClassName() { return getProgram().getClass().getSimpleName(); }
+    public final String getTestedToolName() { return getProgram().getClass().getSimpleName(); }
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java]: Why did you make this change in a few places?

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java [hunk 6]
```diff
@@ -19,6 +19,9 @@
  * This class is an extension of SamFileTester used to test AbstractMarkDuplicatesCommandLineProgram's with SAM files generated on the fly.
  * This performs the underlying tests defined by classes such as AbstractMarkDuplicatesCommandLineProgramTest.
  */
+// TODO: is this really necessary for the packaged testing framework?
+// TODO: it looks like that this is a SamFileTester exclusive for MarkDuplicates with the parameter in GATK/Picard
+// TODO: and thus, it should live in the test resources
 public abstract class AbstractMarkDuplicatesTester extends SamFileTester {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java]: I agree, though I would leave it for a separate pull request to make that change for the testers more broadly.

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java (hunk 2 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: More specifically, refactor the other methods in this class to take the reference dictionary as an argument and have the REFERENCE_FILE just be a string pointer to the file. Utils classes should not have state.

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java (hunk 1 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: This can be overridden, but I think that it is a good practice to have a standard folder structure for test data directories. In addition, this is not project specific, because the java source structure (`src/test/resources`) is more or less an standard.

No action for this comment.

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java (hunk 2 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: This was before an utility class to create `SimpleIntervals` with the reference in the exome test source. It is true that it is not a utils class, but a factory/builder class.

I think that I prefer in this case to hold the dictionary and the refererence to be sure that it correspond to the same one. I renamed to `SimpleIntervalTestFactory` to be clear that it keeps state and it is a factory.

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java (hunk 3 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: I removed the TODO and call the `createInterval(contig, start, start)` instead, to be sure that it is using the sequence dictionary.

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java (hunk 6 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java]: Changed the TODO to a clear procedure and open a new ticket for move it to the test sources (#3762)

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java (hunk 5 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/testers/AbstractMarkDuplicatesTester.java]: This classes were implementing `CommandLineProgramTest`, which is the `CommandLineProgramTester` implementation for GATK (extending also `GATKBaseTest`).

Because this classes are in the testing framework, and they are only tester classes, they do not inherit from `BaseTest` anymore. The method for returning the tool name is called `getTestedToolName` in `CommandLineProgramTester` and thus here it should be the one overridden.

Thus, this is one of the results of finally separating the gatk-testing framework (for re-use with downstream projects) and the test from GATK.

File: src/main/java/org/broadinstitute/hellbender/utils/test/testers/CleanSamIntegrationTest.java (hunk 4 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/testers/CleanSamIntegrationTest.java]: Thank you!

**Comment by magicDGS** [comment]
@jamesemery - we should get this merge as soon as possible to avoid conflicts that pop up in every round of comments. Once this is in, I can go to the open PRs to point out the conflicts and the new structure (e.g., change the new tests to extend `GATKBaseTest`).

I added a new commit addressing the issues and I will rebase to resolve conflicts again.

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java (hunk 2 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: This seems like a good solution, you should update the documentation on the file to reflect this change and make it clear to future authors what this class is doing.

**Comment by jamesemery** [comment]
@magicDGS You need to resolve the conflicts yet again and respond to the comments I made about the `SimpleIntervalTestFactory` then this could probably be merged

File: src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java (hunk 2 above)
Comment (magicDGS) [reply src/main/java/org/broadinstitute/hellbender/utils/test/TargetsToolsTestUtils.java]: Can we left this to a different PR? The main problem with this issue is the `BaseTest` and `GATKBaseTest`, and removing this class by adding new methods to `IntervalUtils` will delay even more the reviewing process...

I open an issue to remove it (https://github.com/broadinstitute/gatk/issues/3771) and I promise to prepare a PR for it once this is in...

**Comment by magicDGS** [comment]
@jamesemery - I rebased to solve the conflict and I open an issue regarding the `SimpleIntervalTestFactory` to do not block this work any longer, because this is already a big change in the testing framework. 

Because the changes are big and we are working in different timezones, conflicts pop up everyday when another PR is accepted before this if they modify any of the test files (which is often the case). If we continue delaying this, it would never be possible to merge...

**Comment by jamesemery** [comment]
@magicDGS Once these conflicts are resolved (which appear to be quick fixes) I will merge this branch.

**Comment by magicDGS** [comment]
Thanks @jamesemery - that's the complication of this big PR. I hope that the tests pass after my last commit and that we can get this in before another PR gets in.

Thanks a lot for reviewing!

**Comment by lbergelson** [comment]
@magicDGS merged!  Thank you for your this.

**Comment by magicDGS** [comment]
Thanks to all of you!

**Comment by magicDGS** [comment]
I tried to warn in every affected open PR as promised @jamesemery
